crates/ra_syntax/src/syntax_error.rs

```diff
@@ -1,209 +1,47 @@
-//! FIXME: write short doc here
+//! Module that defines `SyntaxError`.
 
 use std::fmt;
 
-use ra_parser::ParseError;
-
-use crate::{validation::EscapeError, TextRange, TextUnit, TokenizeError};
+use crate::{TextRange, TextUnit};
 
+/// Represents the result of unsuccessful tokenization, parsing
+/// or semmantical analyzis.
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct SyntaxError {
-    kind: SyntaxErrorKind,
-    location: Location,
-}
-
-// FIXME: Location should be just `Location(TextRange)`
-// TextUnit enum member just unnecessarily compicates things,
-// we should'n treat it specially, it just as a `TextRange { start: x, end: x + 1 }`
-// see `location_to_range()` in ra_ide/src/diagnostics
-#[derive(Clone, PartialEq, Eq, Hash)]
-pub enum Location {
-    Offset(TextUnit),
-    Range(TextRange),
-}
-
-impl From<TextUnit> for Location {
-    fn from(offset: TextUnit) -> Location {
-        Location::Offset(offset)
-    }
-}
-
-impl From<TextRange> for Location {
-    fn from(range: TextRange) -> Location {
-        Location::Range(range)
-    }
-}
-
-impl fmt::Debug for Location {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self {
-            Location::Offset(it) => fmt::Debug::fmt(it, f),
-            Location::Range(it) => fmt::Debug::fmt(it, f),
-        }
-    }
-}
+pub struct SyntaxError(String, TextRange);
+
+// FIXME: there was an unused SyntaxErrorKind previously (before this enum was removed)
+// It was introduced in this PR: https://github.com/rust-analyzer/rust-analyzer/pull/846/files#diff-827da9b03b8f9faa1bade5cdd44d5dafR95
+// but it was not removed by a mistake.
+//
+// So, we need to find a place where to stick validation for attributes in match clauses.
+// Code before refactor:
+// InvalidMatchInnerAttr => {
+//    write!(f, "Inner attributes are only allowed directly after the opening brace of the match expression")
+// }
 
 impl SyntaxError {
-    pub fn new<L: Into<Location>>(kind: SyntaxErrorKind, loc: L) -> SyntaxError {
-        SyntaxError { kind, location: loc.into() }
+    pub fn new(message: impl Into<String>, range: TextRange) -> Self {
+        Self(message.into(), range)
     }
-
-    pub fn kind(&self) -> SyntaxErrorKind {
-        self.kind.clone()
+    pub fn new_at_offset(message: impl Into<String>, offset: TextUnit) -> Self {
+        Self(message.into(), TextRange::offset_len(offset, 1.into()))
```

Comment:
```suggestion
        Self(message.into(), TextRange::offset_len(offset, 0.into()))
```

Reply:
Hmm, I am not sure,  we had `fn location_to_range()` that converted `TextUnit` to `TextRange` of length `1`.
https://github.com/rust-analyzer/rust-analyzer/blob/ea9d18ba836a7228f7310e1bc77c0918f0191a42/crates/ra_ide/src/diagnostics.rs#L118-L123

I don't have good reasons to choose 1 or 0 here, and do you?

Reply:
Or maybe I have one, if we leave the length 1, the we should get error squiggles for one character and if we go with 0, won't we have any? Anyway I need to check this ...

Reply:
With 1, this can be a range outside of the length of the file. The range that represents a position should have length equal to zero. 

When we display such a range to the user we need to translate it to something non-empty, but semantically it should be an empty range.

Reply:
Even with length 0 the same range would be at offset outside of the length of file. Though, I'll trust you in this case.

Reply:
*Just* passt  the  end is ok: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bc7c4e18586ed54c30f8d2cde25de19f

Reply:
Apparently you are right, this is what I found in [vscode docs](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#diagnostics-tips-and-tricks)
> If the start and end positions are the same, VS Code will underline with a squiggle the word at that position.